### ISSUE TITLE
Jaeger Operator with Cassandra create Schema

### ISSUE THREAD

**Issue by sundar-cs** [jaegertracing/jaeger-operator#1147]
I want to try Jaeger streaming strategy with Cassandra (it's running well with Elastic) but to start with I was trying the basic samples based on https://www.jaegertracing.io/docs/1.18/operator/

```yaml
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: cassandra-test1
spec:
  strategy: allInOne
  storage:
    type: cassandra
    cassandraCreateSchema:
      enabled: true
```

Also specifying above with createSchema options:

```yaml
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: cassandra-test2
spec:
  strategy: allInOne # <1>
  allInOne:
    options:
      log-level: debug
  storage:
    type: cassandra
    options: # <2>
      cassandra:
        servers: cassandra
        keyspace: jaeger_v1_datacenter3
    cassandraCreateSchema: # <3>
      datacenter: "datacenter3"
      mode: "prod"
```

In both cases the schema never gets created and the logs show this repeatedly every second forever:

```
Traceback (most recent call last):
  File "/opt/cassandra/bin/cqlsh.py", line 2443, in <module>
    main(*read_options(sys.argv[1:], os.environ))
  File "/opt/cassandra/bin/cqlsh.py", line 2421, in main
    encoding=options.encoding)
  File "/opt/cassandra/bin/cqlsh.py", line 485, in __init__
    load_balancing_policy=WhiteListRoundRobinPolicy([self.hostname]),
  File "/opt/cassandra/bin/../lib/cassandra-driver-internal-only-3.11.0-bb96859b.zip/cassandra-driver-3.11.0-bb96859b/cassandra/policies.py", line 417, in __init__
socket.gaierror: [Errno -2] Name or service not known
Timed out waiting for Cassandra.
```

I have Jaeger Operator, also with debug log enabled (args: ["start", "--log-level=debug"])
But I don't see anything much there except for "Cassandra hostname not specified. Using 'cassandra' for the cassandra-create-schema job." etc for case #1 above.

Please let me know what I'm missing here ?

**Comment by sundar-cs** [comment]
An additional observation is that I was able to get the above along with the Streaming strategy to work well with an existing Cassandra instance.
(the only changes from above pasted yaml are: strategy as streaming, added enabled: false for the schema, and gave existing datacenter and keyspace)

However it would be nice to get the automatic schema creation too to work. So please do let me know.

**Comment by jpkrohling** [comment]
Could you please provide details about your Cassandra installation? The operator expects a Cassandra cluster to exist backed by a service named `cassandra` (it won't provision one for you). This is what the operator expects to exist:

```console
$ kubectl get service cassandra
NAME        TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                                        AGE
cassandra   ClusterIP   None         <none>        7000/TCP,7001/TCP,7199/TCP,9042/TCP,9160/TCP   5s
```

**Comment by sundar-cs** [comment]
Thank you.
Yes in my 2nd approach above I do have the Cassandra cluster backed by service named cassandra.

admin@ip-172-20-40-150:~$ kubectl get svc cassandra
NAME        TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                                        AGE
cassandra   ClusterIP   None         <none>        7000/TCP,7001/TCP,7199/TCP,9042/TCP,9160/TCP   73m

As per the Jaeger documentation link pasted above:
_**When the storage type is set to Cassandra, the operator will automatically create a batch job that creates the required schema for Jaeger to run. This batch job will block the Jaeger installation, so that it starts only after the schema is successfuly created.**_

Does this just mean that in the existing cluster it will create a new data center and keyspace with auto-generated names ?
(If so do we retrieve the keyspace directly by querying Cassandra ? May I ask what is the value add from this ?)

**Comment by jpkrohling** [comment]
The keyspace should be created automatically: 
https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/cassandra/schema/v001.cql.tmpl#L6

But the error you are seeing seems to be that the create-schema job can't connect to Cassandra at all:

```
socket.gaierror: [Errno -2] Name or service not known
```

IIRC, the create-schema image is using a base image that allows you to enter it. I'd recommend doing a `kubectl exec` to enter the pod/container and executing `cqlsh` from there, to debug possible connectivity issues you might be having. If you can't run a simple `cqlsh` query from that container, neither can the create-schema script.

For reference, here's the shell script that bootstraps the create-schema script: https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/cassandra/schema/docker.sh

And here are the other parts to the create-schema. You might want to take a look at the `v00x` files:
https://github.com/jaegertracing/jaeger/tree/master/plugin/storage/cassandra/schema

**Comment by sundar-cs** [comment]
Thank you for your remarks and the links for the scripts.
Yes looks like it's functioning as designed i.e. in my case (#1 above) I did Not have an existing Cassandra cluster.
(whereas in #2 above I had a cluster and gave the datacenter and keyspace names it had, to the Jaeger YAML)

So it explains why I saw the failure (Cassandra cluster did not exist and it will Not get created automatically).
Thanks a lot for clarifying.
